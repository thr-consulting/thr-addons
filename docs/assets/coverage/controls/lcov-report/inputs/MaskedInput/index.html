<!doctype html>
<html lang="en">
<head>
    <title>Code coverage report for inputs/MaskedInput</title>
    <meta charset="utf-8" />
    <link rel="stylesheet" href="../../prettify.css" />
    <link rel="stylesheet" href="../../base.css" />
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <style type='text/css'>
        .coverage-summary .sorter {
            background-image: url(../../sort-arrow-sprite.png);
        }
    </style>
</head>
<body>
<div class='wrapper'>
  <div class='pad1'>
    <h1>
      <a href="../../index.html">All files</a> inputs/MaskedInput
    </h1>
    <div class='clearfix'>
      <div class='fl pad1y space-right2'>
        <span class="strong">22.22% </span>
        <span class="quiet">Statements</span>
        <span class='fraction'>6/27</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">7.41% </span>
        <span class="quiet">Branches</span>
        <span class='fraction'>2/27</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">30% </span>
        <span class="quiet">Functions</span>
        <span class='fraction'>3/10</span>
      </div>
      <div class='fl pad1y space-right2'>
        <span class="strong">26.09% </span>
        <span class="quiet">Lines</span>
        <span class='fraction'>6/23</span>
      </div>
    </div>
    <p class="quiet">
      Press <em>n</em> or <em>j</em> to go to the next uncovered block, <em>b</em>, <em>p</em> or <em>k</em> for the previous block.
    </p>
  </div>
  <div class='status-line low'></div>
<div class="pad1">
<table class="coverage-summary">
<thead>
<tr>
   <th data-col="file" data-fmt="html" data-html="true" class="file">File</th>
   <th data-col="pic" data-type="number" data-fmt="html" data-html="true" class="pic"></th>
   <th data-col="statements" data-type="number" data-fmt="pct" class="pct">Statements</th>
   <th data-col="statements_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="branches" data-type="number" data-fmt="pct" class="pct">Branches</th>
   <th data-col="branches_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="functions" data-type="number" data-fmt="pct" class="pct">Functions</th>
   <th data-col="functions_raw" data-type="number" data-fmt="html" class="abs"></th>
   <th data-col="lines" data-type="number" data-fmt="pct" class="pct">Lines</th>
   <th data-col="lines_raw" data-type="number" data-fmt="html" class="abs"></th>
</tr>
</thead>
<tbody><tr>
	<td class="file low" data-value="MaskedInput.js"><a href="MaskedInput.js.html">MaskedInput.js</a></td>
	<td data-value="22.22" class="pic low"><div class="chart"><div class="cover-fill" style="width: 22%;"></div><div class="cover-empty" style="width:78%;"></div></div></td>
	<td data-value="22.22" class="pct low">22.22%</td>
	<td data-value="27" class="abs low">6/27</td>
	<td data-value="7.41" class="pct low">7.41%</td>
	<td data-value="27" class="abs low">2/27</td>
	<td data-value="30" class="pct low">30%</td>
	<td data-value="10" class="abs low">3/10</td>
	<td data-value="26.09" class="pct low">26.09%</td>
	<td data-value="23" class="abs low">6/23</td>
	</tr>

<tr>
	<td class="file empty" data-value="index.js"><a href="index.js.html">index.js</a></td>
	<td data-value="0" class="pic empty"><div class="chart"><div class="cover-fill" style="width: 0%;"></div><div class="cover-empty" style="width:100%;"></div></div></td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	<td data-value="0" class="pct empty">0%</td>
	<td data-value="0" class="abs empty">0/0</td>
	</tr>

</tbody>
</table>
</div><div class='push'></div><!-- for sticky footer -->
</div><!-- /wrapper -->
<div class='footer quiet pad2 space-top1 center small'>
  Code coverage
  generated by <a href="https://istanbul.js.org/" target="_blank">istanbul</a> at Fri Apr 26 2019 13:53:47 GMT-0500 (CDT)
</div>
</div>
<script src="../../prettify.js"></script>
<script>
window.onload = function () {
        if (typeof prettyPrint === 'function') {
            prettyPrint();
        }
};
</script>
<script src="../../sorter.js"></script>
<script src="../../block-navigation.js"></script>
</body>
</html>
