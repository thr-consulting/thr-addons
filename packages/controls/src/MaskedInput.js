// @flow

import React, {Component} from 'react';
import {Input} from 'semantic-ui-react';
import 'inputmask/dist/inputmask/inputmask.numeric.extensions';
import Inputmask from 'inputmask';

/**
 * See source code for detailed prop types or {@link https://github.com/RobinHerbots/jquery.inputmask|here} for more info.
 * @typedef {Object} inputmaskPropTypes
 */

type Props = {
	value?: string,
	onChange?: (data: string) => {},
	mask?: {
		placeholder?: string,
		optionalmarker?: {
			start: string,
			end: string,
		},
		quantifiermarker?: {
			start: string,
			end: string,
		},
		groupmarker?: {
			start: string,
			end: string,
		},
		alternatormarker?: string,
		escapeChar?: string,
		mask: string,
		repeat?: number,
		greedy?: boolean,
		autoUnmask?: boolean,
		removeMaskOnSubmit?: boolean,
		clearMaskOnLostFocus?: boolean,
		insertMode?: boolean,
		clearIncomplete?: boolean,
		alias?: string,
		onKeyDown?: (event: any, buffer: any, caretPos: number, opts?: Object) => {},
		onBeforeMask?: (initialValue: string, opts?: Object) => string,
		onBeforePaste?: (pastedValue: string, opts?: Object) => string,
		onBeforeWrite?: (event: any, buffer: any, caretPos: number, opts?: Object) => Object,
		onUnMask?: (maskedValue: string, unmaskedValue: string) => string,
		showMaskOnFocus?: boolean,
		showMaskOnHover?: boolean,
		onKeyValidation?: (key: any, result: any) => {},
		numericInput?: boolean,
		rightAlign?: boolean,
		undoOnEscape?: boolean,
		radixPoint?: string,
		groupSeparator?: string,
		keepStatic?: boolean,
		positionCaretOnTab?: boolean,
		tabThrough?: boolean,
		isComplete?: (buffer: any, opts?: Object) => boolean,
		canClearPosition?: (maskset: any, position: number, lastValidPosition: number, opts?: Object) => boolean,
		postValidation?: (buffer: any, currentResult: any, opts?: any) => boolean | Object,
		preValidation?: (buffer: any, pos: number, char: any, isSelection: boolean, opts?: Object) => boolean | Object,
		staticDefinitionSymbol?: string,
		nullable?: boolean,
		positionCaretOnClick?: string,
		casing?: string,
		inputmode?: string,
		colorMask?: string,
	}
};

/**
 * Displays a masked input form. Warning: this component uses jquery for masking so it renders quite slow. Do not use
 * hundreds of these on one screen at the same time.
 * @class
 * @property {string|number} [value=null] - The value to display
 * @property {onChange} [onChange=null] - Called when the value changes.
 * @property {inputmaskPropTypes} [mask=null] - The mask object specified at {@link https://github.com/RobinHerbots/jquery.inputmask|here}.
 */
export default class MaskedInput extends Component<Props> {
	static defaultProps = {
		value: null,
		onChange: null,
		mask: null,
	};

	componentDidMount() {
		if (this.props.value && this._input) this._input.value = this.props.value;
		const im = new Inputmask({
			...this.props.mask,
			oncomplete: this.handleComplete,
			oncleared: this.handleCleared,
			onincomplete: this.handleIncomplete,
		});
		im.mask(this._input);
	}

	componentDidUpdate(prevProps: Object) {
		if (prevProps.value !== this.props.value && this._input) {
			this._input.value = this.props.value || '';
		}
		if (prevProps.mask !== this.props.mask) {
			const im = new Inputmask({
				...this.props.mask,
				oncomplete: this.handleComplete,
				oncleared: this.handleCleared,
				onincomplete: this.handleIncomplete,
			});
			im.mask(this._input);
		}
	}

	props: Props;
	_input: ?HTMLInputElement;

	handleComplete = (ev: Object) => {
		if (this.props.onChange) this.props.onChange(ev.target.value);
	}

	handleCleared = () => {
		if (this.props.onChange) this.props.onChange('');
	}

	handleIncomplete = (ev: Object) => {
		if (this.props.onChange) this.props.onChange(ev.target.value);
	}

	render() {
		const {value, onChange, mask, ...rest} = this.props;
		return (
			<Input {...rest}>
				<input ref={r => (this._input = r)}/>
			</Input>
		);
	}
}
