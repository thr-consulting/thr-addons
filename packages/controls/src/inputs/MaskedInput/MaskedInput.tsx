import debug from 'debug';
import React, {useRef, useEffect} from 'react';
import {Input, InputProps} from 'semantic-ui-react';
import 'inputmask/dist/inputmask/inputmask.numeric.extensions';
import Inputmask from 'inputmask';
import useDeepCompareEffect from 'use-deep-compare-effect';

const d = debug('thx.controls.MaskedInput');

/**
 * See source code for detailed prop types or {@link https://github.com/RobinHerbots/jquery.inputmask|here} for more info.
 * @typedef {Object} inputmaskPropTypes
 */

export interface MaskedInputProps {
	value?: string;
	onChange?: (value?: string) => void;
	onBlur?: () => void;
	mask?: Inputmask.Options;
}

export function MaskedInput(props: MaskedInputProps & Omit<InputProps, 'onChange'>) {
	const {value, onChange, onBlur, mask, ...rest} = props;

	const inputElement = useRef<HTMLInputElement | null>(null);
	const maskInstance = useRef<Inputmask.Instance | null>(null);

	useDeepCompareEffect(() => {
		if (!inputElement.current) throw new Error('Could not get input element');

		d('Creating input mask instance');
		maskInstance.current = new Inputmask({
			...mask,
			oncomplete() {
				if (onChange) onChange(inputElement.current?.value);
				if (mask?.oncomplete) mask.oncomplete();
			},
			oncleared() {
				if (onChange) onChange();
				if (mask?.oncleared) mask.oncleared();
			},
			onincomplete() {
				if (onChange) onChange(inputElement.current?.value);
				if (mask?.onincomplete) mask.onincomplete();
			},
		});
		maskInstance.current.mask(inputElement.current);

		return () => {
			if (maskInstance.current) {
				d('Cleaning up input mask instance');
				maskInstance.current.remove();
				maskInstance.current = null;
			}
		};
	}, [mask]);

	// If we change the value prop we need to sync the DOM value to display the new value
	useEffect(() => {
		if (inputElement.current?.value !== value && value !== undefined) {
			d('Value is changing:', value);
			inputElement.current.value = value;
		}
	}, [value]);

	return (
		<Input {...rest}>
			<input ref={inputElement} onBlur={onBlur} />
		</Input>
	);
}
