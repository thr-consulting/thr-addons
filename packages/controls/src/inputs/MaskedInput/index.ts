export {MaskedInput} from './MaskedInput';
export type {MaskedInputProps} from './MaskedInput';
