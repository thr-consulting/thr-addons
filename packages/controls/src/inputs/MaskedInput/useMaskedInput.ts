import debug from 'debug';
import Inputmask from 'inputmask';
import {useEffect, useRef} from 'react';
import useDeepCompareEffect from 'use-deep-compare-effect';

const d = debug('thx.controls.inputs.MaskedInput.useMaskedInput');

export interface UseMaskedInputProps {
	value?: string;
	onChange?: (value?: string) => void;
	mask?: Inputmask.Options;
	onSet?: (value?: string) => void;
}

export function useMaskedInput(props: UseMaskedInputProps) {
	const {value, onChange, onSet, mask} = props;

	const inputElement = useRef<HTMLInputElement | null>(null);
	const maskInstance = useRef<Inputmask.Instance | null>(null);

	useDeepCompareEffect(() => {
		if (!inputElement.current) throw new Error('Could not get input element');

		d('Creating input mask instance');
		maskInstance.current = new Inputmask({
			...mask,
			oncomplete() {
				if (onChange) onChange(inputElement.current?.value);
				if (mask?.oncomplete) mask.oncomplete();
			},
			oncleared() {
				if (onChange) onChange();
				if (mask?.oncleared) mask.oncleared();
			},
			onincomplete() {
				if (onChange) onChange(inputElement.current?.value);
				if (mask?.onincomplete) mask.onincomplete();
			},
		});
		maskInstance.current.mask(inputElement.current);

		return () => {
			if (maskInstance.current) {
				d('Cleaning up input mask instance');
				maskInstance.current.remove();
				maskInstance.current = null;
			}
		};
	}, [mask]);

	// If we change the value prop we need to sync the DOM value to display the new value
	useEffect(() => {
		if (inputElement.current && inputElement.current?.value !== value && value !== undefined) {
			d('Value is changing:', value);
			inputElement.current.value = value;
			onSet && onSet(value);
		}
	}, [value]);

	return inputElement;
}
