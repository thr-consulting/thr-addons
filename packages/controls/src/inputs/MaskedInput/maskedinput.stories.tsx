import debug from 'debug';
import React, {useState} from 'react';
import {Form, Container, Button, Segment, Input} from 'semantic-ui-react';
import {MaskedInput} from './MaskedInput';

const d = debug('thx.controls.MaskedInput.stories');

export default {title: 'MaskedInput'};

export const Main = () => {
	const [shown, setShown] = useState(true);
	const [value, setValue] = useState();
	const [mask, setMask] = useState('99-999-99');
	const [maskTemp, setMaskTemp] = useState('99-999-99');

	if (!shown) {
		return (
			<Container>
				<Button onClick={() => setShown(true)}>Show</Button>
			</Container>
		);
	}

	return (
		<Container>
			<Segment basic>
				<Form>
					<Form.Field inline width={6}>
						<label>Masked Input</label>
						<MaskedInput
							mask={{
								mask,
								autoUnmask: true,
								showMaskOnHover: false,
							}}
							onChange={v => {
								d('onChange', v);
								setValue(v);
							}}
							onBlur={() => {
								d('onBlur');
							}}
							value={value}
						/>
					</Form.Field>
				</Form>
			</Segment>
			<Segment basic>
				<Input
					action={{
						content: 'Set Mask',
						onClick: () => {
							setMask(maskTemp);
						},
					}}
					value={maskTemp}
					onChange={v => setMaskTemp(v.currentTarget.value)}
				/>
			</Segment>
			<Segment basic>
				<Button
					onClick={() => {
						setValue('8811245');
					}}
				>
					Set value
				</Button>
				<Button onClick={() => setShown(false)}>Hide</Button>
			</Segment>
			<Segment>Value is: {value}</Segment>
		</Container>
	);
};
