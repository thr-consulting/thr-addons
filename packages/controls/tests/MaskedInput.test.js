import React from 'react';
import {shallow} from 'enzyme';
import MaskedInput from '../src/MaskedInput';

describe('MaskedInput', () => {
	it('should render without throwing an error', () => {
		expect(shallow(<MaskedInput mask={{mask: '99-999-99', autoUnmask: true, showMaskOnHover: false}}/>)).toMatchSnapshot();
	});
});
