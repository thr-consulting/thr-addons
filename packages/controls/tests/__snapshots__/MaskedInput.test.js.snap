// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaskedInput should render without throwing an error 1`] = `
<Input
  type="text"
>
  <input />
</Input>
`;
